File: gappers.py
import random
from collections import Counter

# Define the deck of cards
suits = ['hearts', 'diamonds', 'clubs', 'spades']
ranks = list(range(2, 15))  # 2-10, Jack=11, Queen=12, King=13, Ace=14
deck = [(rank, suit) for suit in suits for rank in ranks]

def is_flush(hand, player_suit):
    suit_counts = Counter(suit for rank, suit in hand if suit == player_suit)
    return max(suit_counts.values()) >= 5

def is_gapper(hand):
    suited_cards = [rank for rank, suit in hand if suit == hand[0][1]]
    suited_cards.sort()
    for i in range(len(suited_cards) - 2):
        if suited_cards[i+2] - suited_cards[i] == 2:
            return True
    return False

def simulate_hand(hand_type):
    random.shuffle(deck)
    if hand_type == 'gapper':
        suit = random.choice(suits)
        middle_rank = random.choice(ranks[1:-1])  # Exclude the lowest and highest ranks to have room for a gapper
        player_hand = [(middle_rank - 1, suit), (middle_rank + 1, suit)]
    else:  # any two suited cards
        suit = random.choice(suits)
        player_hand = [(rank, suit) for rank in random.sample(ranks, 2)]
    community_cards = [card for card in deck if card not in player_hand][:5]
    all_cards = player_hand + community_cards
    return is_flush(all_cards, suit)

def estimate_probabilities(n=1_000_000):
    gapper_flush_counts = 0
    suited_flush_counts = 0
    for _ in range(n):
        if simulate_hand('gapper'):
            gapper_flush_counts += 1
        if simulate_hand('suited'):
            suited_flush_counts += 1
    return gapper_flush_counts / n, suited_flush_counts / n

gapper_flush_prob, suited_flush_prob = estimate_probabilities()
print(f"Gapper flush probability: {gapper_flush_prob}")
print(f"Suited flush probability: {suited_flush_prob}")

File: gappers.rs
use rand::seq::SliceRandom;
use rand::Rng;

#[derive(Clone, Copy, PartialEq, Eq)]
enum Rank {
    Two, Three, Four, Five, Six, Seven, Eight, Nine, Ten, Jack, Queen, King, Ace,
}

#[derive(Clone, Copy, PartialEq, Eq)]
enum Suit {
    Hearts, Diamonds, Clubs, Spades,
}

enum HandType {
    Gapper,
    Suited,
}

fn is_flush(hand: &Vec<(Rank, Suit)>, suit: &Suit) -> bool {
    hand.iter().filter(|&card| card.1 == *suit).count() >= 5
}

fn simulate_hand(hand_type: HandType) -> bool {
    let mut rng = rand::thread_rng();
    let suits = vec![Suit::Hearts, Suit::Diamonds, Suit::Clubs, Suit::Spades];
    let ranks: Vec<Rank> = vec![Rank::Two, Rank::Three, Rank::Four, Rank::Five, Rank::Six, Rank::Seven, Rank::Eight, Rank::Nine, Rank::Ten, Rank::Jack, Rank::Queen, Rank::King, Rank::Ace];
    let mut deck: Vec<(Rank, Suit)> = ranks.iter().flat_map(|&rank| suits.iter().map(move |&suit| (rank, suit))).collect();
    deck.shuffle(&mut rng);
    let suit = suits.choose(&mut rng).unwrap();
    let player_hand;
    match hand_type {
        HandType::Gapper => {
            let middle_rank_index = rng.gen_range(1..ranks.len() - 1);
            player_hand = vec![(ranks[middle_rank_index - 1], *suit), (ranks[middle_rank_index + 1], *suit)];
        },
        HandType::Suited => {
            let player_ranks: Vec<Rank> = ranks.choose_multiple(&mut rng, 2).cloned().collect();
            player_hand = player_ranks.into_iter().map(|rank| (rank, *suit)).collect();
        },
    }
    let community_cards: Vec<(Rank, Suit)> = deck.into_iter().filter(|card| !player_hand.contains(card)).take(5).collect();
    let mut all_cards = player_hand.clone();
    all_cards.extend(community_cards);
    is_flush(&all_cards, suit)
}

fn estimate_probabilities(n: i32) -> (f64, f64) {
    let mut gapper_flush_counts = 0;
    let mut suited_flush_counts = 0;
    for _ in 0..n {
        if simulate_hand(HandType::Gapper) {
            gapper_flush_counts += 1;
        }
        if simulate_hand(HandType::Suited) {
            suited_flush_counts += 1;
        }
    }
    (gapper_flush_counts as f64 / n as f64, suited_flush_counts as f64 / n as f64)
}

fn main() {
    let (gapper_flush_prob, suited_flush_prob) = estimate_probabilities(100_000_000);
    println!("Gapper flush probability: {}", gapper_flush_prob);
    println!("Suited flush probability: {}", suited_flush_prob);
}
